client/frontend/client.go

```diff
@@ -21,17 +21,561 @@
 package frontend
 
 import (
-	"github.com/uber/cadence/.gen/go/cadence/workflowserviceclient"
-	"github.com/uber/cadence/common"
+	"context"
+	"sync"
+	"time"
+
 	"go.uber.org/yarpc"
+
+	"github.com/pborman/uuid"
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/membership"
+	"go.uber.org/cadence/.gen/go/cadence/workflowserviceclient"
+	"go.uber.org/cadence/.gen/go/shared"
 )
 
-// Client is the interface exposed by frontend service client
-type Client interface {
-	workflowserviceclient.Interface
+var _ Client = (*clientImpl)(nil)
+
+type clientImpl struct {
+	resolver        membership.ServiceResolver
+	thriftCacheLock sync.RWMutex
+	thriftCache     map[string]workflowserviceclient.Interface
+	rpcFactory      common.RPCFactory
+}
+
+// NewClient creates a new frontend service TChannel client
+func NewClient(d common.RPCFactory, monitor membership.Monitor) (Client, error) {
```

Comment:
Can you also wrap this with a metricClient to emit client-side metric on the caller.  Similar pattern we do for other clients.